dewy/common/db_migration.py

```diff
@@ -0,0 +1,120 @@
+import hashlib
+import os
+import re
+from typing import Dict, Iterable, Optional, Tuple
+
+import asyncpg
+from loguru import logger
+
+
+async def apply_migrations(
+    conn: asyncpg.Connection, *, migration_dir: str = "migrations"
+):
+    """Applies the migrations to the database."""
+
+    # TODO: For deployments, we may wish to make this a script that
+    # can be executed, rather than always calling it on startup.
+
+    # Ensure the migration table exists.
+    await conn.execute(
+        """
+        CREATE TABLE IF NOT EXISTS migration (
+            -- Record the versions we have applied.
+            id INTEGER NOT NULL,
+            -- Record the SHA of the version so we can detect changes.
+            sha256 TEXT,
+            PRIMARY KEY (id)
+        );
+        """
+    )
+
+    defined_migrations = _get_defined_migrations(migration_dir)
+    applied_migrations = await _get_applied_migrations(conn)
+
+    applied = 0
+    for migration_id, migration_file in defined_migrations:
+        applied_sha256 = applied_migrations.pop(migration_id, None)
```

Comment:
This seems like it could create problems if a migration file changes after being applied. The new migration would be applied on top of the old one. I'm not sure if that's better or worse than not applying the migration, but most of the tools I've used in the past use logical identifiers (ie the filename) for this logic.

Reply:
No -- this isn't used to skip applying the migration. It actually causes it to raise an error if the content of a given migration has changed. The intention here is to detect cases where we change the migration (because we're still in early development) and allow us to just raise an exception saying "you need to recreate the DB in this scenario".

The hash is used as a fingerprint / checksum to make sure that if a DB thinks it has applied the migration, that it has actually applied *the same migration*.